src/lib/tree/data-source/flat-data-source.ts

```diff
@@ -0,0 +1,153 @@
+ *     key: 'Fruits',
+ *     children: [
+ *       NodeOne: {
+ *         key: 'Apple',
+ *       },
+ *       NodeTwo: {
+ *        key: 'Pear',
+ *      }
+ *    ]
+ *  }
+ *  After flattener flatten the tree, the structure will become
+ *  SomeNode: {
+ *    key: 'Fruits',
+ *    expandable: true,
+ *    level: 1
+ *  },
+ *  NodeOne: {
+ *    key: 'Apple',
+ *    expandable: false,
+ *    level: 2
+ *  },
+ *  NodeTwo: {
+  *   key: 'Pear',
+  *   expandable: false,
+  *   level: 2
+  * }
+ */
+export class MatTreeFlattener<T, F> {
+
+  constructor(public transformFunction: (node: T, level: number) => F,
+              public getLevel: (node: F) => number,
+              public isExpandable: (node: F) => boolean,
+              public getChildren: (node: T) => Observable<T[]>) {}
+
+  _flattenNode(node: T, level: number,
+               resultNodes: F[], parentMap: boolean[]): F[] {
+    const flatNode = this.transformFunction(node, level);
+    resultNodes.push(flatNode);
+
+    if (this.isExpandable(flatNode)) {
+      this.getChildren(node).pipe(take(1)).subscribe(children => {
+        children.forEach((child, index) => {
+          let childParentMap: boolean[] = parentMap.slice();
+          childParentMap.push(index != children.length - 1);
+          this._flattenNode(child, level + 1, resultNodes, childParentMap);
+        });
+      });
+    }
+    return resultNodes;
+  }
+
+  /**
+   * Flatten a list of node type T to flattened version of node F.
+   * Please note that type T may be nested, and the length of `structuredData` may be different
+   * from that of returned list `F[]`.
+   */
+  flattenNodes(structuredData: T[]): F[] {
+    let resultNodes: F[] = [];
+    structuredData.forEach(node => this._flattenNode(node, 0, resultNodes, []));
+    return resultNodes;
+  }
+
+  /**
+   * Expand flattened node with current expansion status.
+   * The returned list may have different length.
+   */
+  expandFlattenedNodes(nodes: F[], treeControl: TreeControl<F>): F[] {
+    let results: F[] = [];
+    let currentExpand: boolean[] = [];
+    currentExpand[0] = true;
+
+    nodes.forEach((node) => {
+      let expand = true;
+      for (let i = 0; i <= this.getLevel(node); i++) {
+        expand = expand && currentExpand[i];
+      }
+      if (expand) {
+        results.push(node);
+      }
+      if (this.isExpandable(node)) {
+        currentExpand[this.getLevel(node) + 1] = treeControl.isExpanded(node);
+      }
+    });
+    return results;
+  }
+}
+
+
+/**
+ * Data source for flat tree.
+ * The data source need to handle expansion/collapsion of the tree node and change the data feed
+ * to `MatTree`.
+ * The nested tree nodes of type `T` are flattened through `MatTreeFlattener`, and converted
+ * to type `F` for `MatTree` to consume.
+ */
+export class MatTreeFlatDataSource<T, F> implements DataSource<F> {
+  _flattenedData = new BehaviorSubject<F[]>([]);
+  get flattenedData() { return this._flattenedData.value; }
```

Comment:
Why expose `flattenedData` in addition to the data stream coming back from `connect`?

Reply:
Removed getters